blog/2023-08-21-soc-vertx-pinot-client.mdx

```diff
@@ -0,0 +1,120 @@
+    development and thus, the finally released versions may differ.
+
+</Alert>
+
+## Introduction
+
+Eclipse Vert.x is a toolkit to build reactive applications on the Java virtual machine. It provides asynchronous and
+non-blocking clients for different types of databases. Apache Pinot is a realtime distributed datastore for analytics
+workloads. The [Apache Pinot Java Client](https://docs.pinot.apache.org/users/clients/java) uses
+[AsyncHttpClient](https://github.com/AsyncHttpClient/async-http-client) as the default transport and is cumbersome to
+use in Eclipse Vert.x applications.
+
+This project proposes to create a vertx-pinot-client extension. This extension will be a lightweight wrapper around
+the existing Pinot Java client. It will also offer alternative transport for the client using vertx-web-client.
+This will make the Apache Pinot more accessible to Vert.x users.
+
+## Usage
+
+Add the pinot-client dependency to your project.
+
+```xml
+<dependency>
+    <groupId>io.reactiverse</groupId>
+    <artifactId>pinot-client</artifactId>
+    <version>1.0-SNAPSHOT</version>
+</dependency>
+```
+
+You can create an instance of the client to interact with the Pinot cluster as follows.
+
+```java
+String brokerUrl = "localhost:8000";
+VertxPinotClientTransport transport = new VertxPinotClientTransport(vertx);
+VertxConnection connection = VertxConnectionFactory.fromHostList(vertx, List.of(brokerUrl), transport);
+```
+
+Now we can use the Connection to query the cluster. Here is an example to query the top 10 home run scorers from
+the sample quickstart dataset.
+
+```java
+String query = "select playerName, sum(homeRuns) AS totalHomeRuns from baseballStats where homeRuns > 0 group by playerID, playerName ORDER BY totalHomeRuns DESC limit 10";
+connection
+    .execute(query)
+    .onSuccess(resultSetGroup -> {
+        ResultSet results = resultSetGroup.getResultSet(0);
+        System.out.println("Player Name\tTotal Home Runs");
+        for (int i = 0; i < results.getRowCount(); i++) {
+            System.out.println(results.getString(i, 0) + "\t" + results.getString(i, 1));
+        }
+    })
+    .onFailure(Throwable::printStackTrace);
+```
+
+## Demo
+
+A more real-life demo is available at [`pizza-shop-demo`](https://github.com/lucifer4j/pizza-shop-demo). It includes a
+pizza delivery service specializing in pizzas with Indian toppings. They have a dashboard built using Apache Pinot
+to monitors their orders in realtime. The dashboard deals with three types of data: products, users, and orders. The
+dashboard uses the Vert.x Pinot Client to [`query`](https://github.com/lucifer4j/pizza-shop-demo/blob/main/dashboard/backend/src/main/java/pizzashop/DashboardVerticle.java)
+the Pinot cluster.
+
+![Order Dashboard Statistics](/images/blog/soc-vertx-pinot-client/demo-dashboard-1.png)
+![Order Dashboard Latest and Popular Items](/images/blog/soc-vertx-pinot-client/demo-dashboard-2.png)
+
+## Challenges
```

Comment:
Perhaps you could also mention the problem with blocking java `Future`? Also, how you solved it temporarily with `executeBlocking`, and the conversation with Pinot commiters to get the pending PR for `CompletableFuture` merged. And how it will remove the need for offloading to the Vert.x worker pool.

Reply:
done